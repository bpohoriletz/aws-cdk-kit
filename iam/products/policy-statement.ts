import * as iam from 'aws-cdk-lib/aws-iam';

export class PolicyStatementProduct implements IPolicyStatementProduct {
  actions: string[];
  resources: string[];

  constructor() {
    this.actions = [];
    this.resources = [];
  }

  addAction(action: string): this {
    this.actions.push(action);

    return this;
  }

  addResource(resource: string): this {
    this.resources.push(resource);

    return this;
  }
}

interface IPolicyStatementProduct extends iam.PolicyDocumentProps {
  actions?: string[];
  resources?: string[];
}

export interface IPolicyStatementBuilder {
  setPermissions(): IPolicyStatementBuilder;
  addResource(resource: string): IPolicyStatementBuilder;
  getResult(): iam.PolicyStatementProps;
}

export interface IPolicyStatementBuilderConstructor {
  new (): IPolicyStatementBuilder;
}
