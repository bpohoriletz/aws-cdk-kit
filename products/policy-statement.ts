import * as iam from 'aws-cdk-lib/aws-iam';

export class PolicyStatementProduct implements iam.PolicyStatementProps {
  actions: string[];
  resources: string[];

  constructor() {
    this.actions = [];
    this.resources = [];
  }

  addAction(action: string): this {
    this.actions.push(action);

    return this;
  }

  addResource(resource: string): this {
    this.resources.push(resource);

    return this;
  }
}

export interface IPolicyStatementBuilder {
  setRead?(): this;
  setWrite?(): this;
  addResource?(resource: string): this;
  getResult(): PolicyStatementProduct;
}

export interface IPolicyStatementBuilderConstructor {
  new (): IPolicyStatementBuilder;
}
