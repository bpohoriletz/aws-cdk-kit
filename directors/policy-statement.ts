import * as iam from 'aws-cdk-lib/aws-iam';
import { IPolicyStatementBuilder, IPolicyStatementBuilderConstructor } from '../products/policy-statement';

export default class PolicyStatementDirector {
  private builder: IPolicyStatementBuilder;

  constructor(builder: IPolicyStatementBuilderConstructor) {
    this.builder = new builder();
  }

  constructPolicyStatement(resources: string[] = ['*']): iam.PolicyStatement {
    resources.forEach((resource) => {
      this.builder.addResource(resource);
    });
    this.builder.setPermissions();

    return new iam.PolicyStatement(this.builder.getResult());
  }
}
