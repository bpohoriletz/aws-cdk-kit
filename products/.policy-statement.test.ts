import { PolicyStatementProduct } from './policy-statement';

describe('new PolicyStatementProduct()', () => {
  let instance: PolicyStatementProduct;

  beforeEach(() => {
    instance = new PolicyStatementProduct();
  });

  test('has correct defaults', () => {
    expect(instance).toMatchSnapshot();
  });
});
