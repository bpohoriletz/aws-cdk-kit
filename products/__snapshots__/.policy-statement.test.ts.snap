// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new PolicyStatementProduct() has correct defaults 1`] = `
PolicyStatementProduct {
  "actions": [],
  "resources": [],
}
`;
